DOCS/man/mpv: improve path docs and clarify config-dir path behavior

Arcitec opened the pull request:
The previous documentation was pretty scary and misleading, and was also outdated in several sections.

It was also difficult to understand some of the paragraphs.

The path documentation has now been overhauled to be easier to understand and to document the latest mpv behavior.

This change documents the fact that there's no difference between `~~/foo` and `~~home/foo` in terms of that "scary bug". They both use the exact same code and have the same `--no-config` "bug" behavior.

References:

- [--no-config](https://github.com/mpv-player/mpv/blob/master/DOCS/man/options.rst#program-behavior)
- [old docs which makes it sound like ~~/ is buggier than ~~home/](https://github.com/mpv-player/mpv/blob/master/DOCS/man/mpv.rst#paths)
- [code proof that both actually behave *identically* when falling back to "mp_get_platform_path"](https://github.com/mpv-player/mpv/blob/0c5420b4122689d0b9f223a554eadffddbf752d7/options/path.c#L183)

Edit: Improved further, as described here: https://github.com/mpv-player/mpv/pull/15218#issuecomment-2448776550

sfan5 reviewed (dismissed):
lgtm

Arcitec commented on DOCS/man/mpv.rst:
```diff
@@ -514,14 +514,17 @@ Currently, the prefix ``~~home/`` expands to the mpv configuration directory
 ``~/`` expands to the user's home directory. (The trailing ``/`` is always
 required.) The following paths are currently recognized:
 
+**Important:** Beware that if ``--no-config`` is used, then all of the "config
```
Oh thanks for pointing that out. I'll try to figure out the admonition syntax now!

Arcitec commented:
@sfan5 @na-na-hi I used this `scripts/testpath.lua` script to discover a few other quirks, and have now rewritten the path documentation to remove a lot of outdated sentences and improve overall clarity. It now uses the admonition style as mentioned by nana.

```lua
local mp = require("mp")
local msg = require("mp.msg")

for _, prefix in ipairs({ "~~global/", "~~home/", "~~/" }) do
    for _, n in ipairs({ "", "foo/bar" }) do
        local p = prefix .. n
        local res = mp.command_native({ "expand-path", p })
	    msg.info(p .. ": " .. type(res) .. " = <" .. tostring(res) .. ">")
    end
end

for _, n in ipairs({ "encoding-profiles.conf", "mpv.conf" }) do
	local res = mp.find_config_file(n)
	msg.info("find: " .. n .. ": " .. type(res) .. " = <" .. tostring(res) .. ">")
end
```

Behavior:

```sh
$ mpv movie.mkv
[testpath] ~~global/: string = </etc/mpv>
[testpath] ~~global/foo/bar: string = </etc/mpv/foo/bar>
[testpath] ~~home/: string = </home/johnny/.config/mpv>
[testpath] ~~home/foo/bar: string = </home/johnny/.config/mpv/foo/bar>
[testpath] ~~/: string = </home/johnny/.config/mpv>
[testpath] ~~/foo/bar: string = </home/johnny/.config/mpv/foo/bar>
[testpath] find: encoding-profiles.conf: string = </etc/mpv/encoding-profiles.conf>
[testpath] find: mpv.conf: string = </home/johnny/.config/mpv/mpv.conf>

$ mpv --no-config --script=/home/johnny/.config/mpv/scripts/testpath.lua movie.mkv
[testpath] ~~global/: string = <>
[testpath] ~~global/foo/bar: string = <foo/bar>
[testpath] ~~home/: string = <>
[testpath] ~~home/foo/bar: string = <foo/bar>
[testpath] ~~/: string = <>
[testpath] ~~/foo/bar: string = <foo/bar>
[testpath] find: encoding-profiles.conf: nil = <nil>
[testpath] find: mpv.conf: nil = <nil>
```

Arcitec commented:
Ah yeah. I've fixed that now.

I don't wrap commit messages normally, since it's actively harmful to the Git history and totally pointless and causes tons of extra work of manually wrapping and rewriting the lines until they wrap nicely (and some lines will then be very wide and others short since certain words won't fit on the same line), and even rephrasing/adding/removing some words later can take minutes of moving the lines around to adapt to their new wrapping. And then when you view the patches in anything that won't exactly fit 72 characters wide, you'll get wonderful "double wrapping" where the viewport also wraps your already manually wrapped lines into a mess of small fragments...

It's a very painful convention that dates back to manually emailed patches and mailing lists.

But when it's already the convention of a project it should of course be followed. Sorry about that! :)

kasper93 commented on DOCS/man/mpv.rst:
```diff
@@ -499,32 +499,53 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
+    empty string, and ``expand-path "~~home/foo/bar"`` will only return ``foo/bar``
+    (without any path prefix), which might not be what you expected.
+
+    Furthermore, certain commands such as ``mp.find_config_file("mpv.conf")``
```
This is internal API, why is this mentioned here? Why external scripts uses it?

Arcitec replied:
I saw it in `lua.c` and in various internal `.lua` scripts and am using it myself to find files in the config-dirs. It does the right thing: Searches all config dirs for the given path.

I did not find any alternatives that can search all config dirs from user scripts.

Arcitec replied:
Anyway, I'll remove the mention of the specific command and just say that any commands that search in config dirs will fail.

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,32 +499,53 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
 following arguments as options.
 
-When using the client API, you should strictly avoid using ``mpv_command_string``
+For paths passed to mpv CLI options, the situation is further complicated by the
```
this is not correct, the old docs intentionally talked about "suboptions", because only then you have a problem with `,` or `:`.
if you have a simple option that takes a single string (path), you can pass anything to it without problems.

Arcitec replied:
Ahh, I will rephrase to match the original explanation.

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,32 +499,53 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
+    Beware that if ``--no-config`` is used, all of the "config directory"-based
+    paths (``~~/``, ``~~home/`` and ``~~global/``) will resolve to empty strings.
+
+    This means that commands such as ``expand-path "~~home/"`` will return an
```
this is a concrete example so "such as" doesn't make sense here.
I'd rather say
> This means that `~~home/` will expand to any empty string, while `~~home/foo/bar` will expand to `foo/bar`
> (a relative path), which can be unexpected.

Arcitec replied:
I wanted to include an example of a command that's common and affected by this, but I'll rephrase it! :)

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,32 +499,53 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
-                 Note that if --no-config is used ``~~/foobar`` will resolve to
-                 ``foobar`` which can be unexpected.
-``~/``           user home directory root (similar to shell, ``$HOME``)
+``~/``           user home directory root (similar to a shell's ``$HOME``)
```
this changes the meaning subtly.
old: similar to how it works in a shell, equivalent to `$HOME`
new: similar to how `$HOME` works in a shell
the old one is more precise

Arcitec replied:
The old one was really imprecise. But I'll find a new phrasing.

Arcitec commented:
Thanks for the feedback. I have pushed resolutions to all comments. The `suboption` example was a big miss (didn't realize it referred to `:` and `,`-separated values, which was due to the lack of clarity in the original paragraph), and the other feedback also lead to even more precise documentation.

Changes:

https://github.com/mpv-player/mpv/compare/4a73cdca17f399d10ae55513280072bb964b183b..d64d1fe9600824e42981dbe377bd77a1425596b3

Arcitec commented on DOCS/man/mpv.rst:
```diff
@@ -499,32 +499,53 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
-                 Note that if --no-config is used ``~~/foobar`` will resolve to
-                 ``foobar`` which can be unexpected.
-``~/``           user home directory root (similar to shell, ``$HOME``)
+``~/``           user home directory root (equivalent to ``$HOME`` in a shell)
```
Yeah good point, we should mention `~/` shell paths as they are the most obvious equivalent. I'll do that now.

Arcitec commented:
Per last review comment:

- Improved clarity of the description of `~/` by mentioning the terminal/shell environment variable equivalents, which are better examples than the old `$HOME` example.
- Moved `~/` to the top of the list so that the rest of the list flows better, and replaced the confusing "user home directory root" with the less confusing "The current user's home directory".
- Fixed the inconsistent use of "Sentence case." and punctuation in the other list entries.

Changes:

https://github.com/mpv-player/mpv/compare/d64d1fe9600824e42981dbe377bd77a1425596b3..f1a92e30686210477046ed18f01c7db5e31d204c

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,41 +499,63 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
 following arguments as options.
 
-When using the client API, you should strictly avoid using ``mpv_command_string``
+For paths passed to mpv CLI suboptions (options that have multiple `:` and
```
nit: this is not specific to CLI
```suggestion
For paths passed to mpv suboptions (options that have multiple `:` and
```

Arcitec replied:
Good point, thanks!

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,41 +499,63 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
-additionally wrapped in the fixed-length syntax, e.g. ``%n%string_of_length_n``
-(see above).
+The same applies when you're using the scripting API, where you should avoid using
+``mp.command``, and instead prefer using the "separate parameters" variants (such
+as ``mp.commandv`` or ``mp.command_native``), so that the filenames will always be
+correctly interpreted.
```
i think it's clear enough without being so verbose
```suggestion
``mp.command``, and instead prefer using ``mp.commandv`` or ``mp.command_native``.
```

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,41 +499,63 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
+    paths (``~~/``, ``~~home/`` and ``~~global/``) will be empty strings.
+
+    This means that ``~~home/`` would expand to an empty string, and that
+    sub-paths such as ``~~home/foo/bar"`` would expand to "relative paths"
+    without any path prefix (``foo/bar``), which may not be what you expected.
```
```suggestion
    sub-paths such as ``~~home/foo/bar"`` would expand to a 
    relative path (``foo/bar``), which may not be what you expected.
```

Arcitec replied:
Good change!

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,41 +499,63 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
+    without any path prefix (``foo/bar``), which may not be what you expected.
+
+    Furthermore, any commands that search in config directories will fail
+    to find anything, since there won't be any directories to search in.
```
honestly not sure if we should state this, because it's so obvious

Arcitec replied:
I was doubtful too, but I figured it's the "Paths" section and that it's worth mentioning what happens to paths when `--no-config` is used. I don't know either. I'll leave it for now but if you have a strong opinion we can remove it next.

sfan5 commented on DOCS/man/mpv.rst:
```diff
@@ -499,41 +499,63 @@ console controls. (Which makes it suitable for playing data piped to stdin.)
 Name             Meaning
 ================ ===============================================================
-``~~/``          If the subpath exists in any of the mpv's config directories
+``~/``           The current user's home directory (equivalent to ``~/`` and
+                 ``$HOME/`` in terminal environments).
```
these two are the same thing
```suggestion
``~/``           The current user's home directory (just like ``$HOME`` in a shell).
```

Arcitec replied:
Yeah, but I was asked to add both examples in a previous review, with the motivation that `~/` in mpv is easier to understand when comparing it to `~/` in a shell rather than `$HOME/`. I was also asked to add the trailing slashes for extra clarity. Hehe. :D

Arcitec commented:
Changes:

https://github.com/mpv-player/mpv/compare/f19903099f0e15ac6d370402ae50605dfd5a0983..86d7425852d2c7ba3da014452ae5626f5994c631

- Indicate that suboptions aren't just coming from the CLI.
- Shorten the scripting API filename safety paragraph.
- Shorten the relative path expansion paragraph in `--no-config` section.
- Came up with a solution for the `exe_dir` phrasing so that it follows the style of the other options.

Hopefully it's ready now! :)

Dudemanguy reviewed (approved):
I didn't follow most of the discussion but the additional clarity in the docs seems good to me.
